### PR TITLE
Merge latest bits from `main` into `roles` feature branch and resolve conflicts around primary constructors

### PR DESCRIPTION
**Pull request by jcouv**
When I did the merge, I didn't address any of the conflicts. All the conflicts are resolved in the final commit separately: ~~https://github.com/dotnet/roslyn/pull/67362/commits/983cc650c011716fbf11d345921cb2b4422bd3b0~~ https://github.com/dotnet/roslyn/commit/2b0adb74b7ccbf5d2905d283c376f6cdb4c2711f

Relates to test plan https://github.com/dotnet/roslyn/issues/66722

### CODE REVIEW

**Review by 333fred — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 1]
```diff
@@ -350,14 +350,8 @@ private static bool IsTypeModifierOrTypeKeyword(SyntaxKind kind)
                 case SyntaxKind.SealedKeyword:
                 case SyntaxKind.StaticKeyword:
                 case SyntaxKind.UnsafeKeyword:
-<<<<<<< HEAD
-                case SyntaxKind.OpenBracketToken:
                 case SyntaxKind.ExplicitKeyword:
                 case SyntaxKind.ImplicitKeyword:
-||||||| e9f1cd29802
-                case SyntaxKind.OpenBracketToken:
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: It looks like ```SyntaxKind.OpenBracketToken``` got dropped. Is this intentional or am I misinterpreting the merge?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: It looks like I misinterpreted the merge

File: src/Compilers/CSharp/Portable/Syntax/Syntax.xml [hunk 2]
```diff
@@ -3515,6 +3514,7 @@
       <Kind Name="IdentifierToken"/>
     </Field>
     <Field Name="TypeParameterList" Type="TypeParameterListSyntax" Optional="true" Override="true"/>
+    <Field Name="ParameterList" Type="ParameterListSyntax" Optional="true" Override="true" />
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Syntax/Syntax.xml:3517]: This doesn't look like a merge. This looks like a feature work and I would prefer that work to be done in a separate PR.

**Review by AlekseyTs — CHANGES_REQUESTED** [review]
It looks like this PR includes non-trivial feature work and, consequently, non-trivial test changes. Also, the PR is too big for CodeFlow to handle. Therefore, it is hard to be confident in the quality of a code review. I think the feature work should be separated into a different PR, and the merge PR should be scoped down as follows:
- Simply getting code compiled and tests, that are not primarily related to the feature, passed.
- Disabling feature tests that start failing, if any, and handling them in a follow-up PR.

File: src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs [hunk 3]
```diff
@@ -2840,15 +2840,27 @@ public static ConversionOperatorMemberCrefSyntax ConversionOperatorMemberCref(Sy
...
-                forUnderlyingType, baseList, constraintClauses,
-                SyntaxFactory.Token(SyntaxKind.OpenBraceToken), members, SyntaxFactory.Token(SyntaxKind.CloseBraceToken), semicolonToken: default);
-||||||| e9f1cd29802
-=======
+        public static ExtensionDeclarationSyntax ExtensionDeclaration(SyntaxList<AttributeListSyntax> attributeLists, SyntaxTokenList modifiers, SyntaxToken implicitOrExplicitKeyword, SyntaxToken identifier, TypeParameterListSyntax? typeParameterList, ParameterListSyntax? parameterList, ForTypeSyntax? forUnderlyingType, BaseListSyntax? baseList, SyntaxList<TypeParameterConstraintClauseSyntax> constraintClauses, SyntaxList<MemberDeclarationSyntax> members)
+            => SyntaxFactory.ExtensionDeclaration(attributeLists, modifiers,
+                implicitOrExplicitKeyword, SyntaxFactory.Token(SyntaxKind.ExtensionKeyword), identifier,
+                typeParameterList, parameterList, forUnderlyingType, baseList, constraintClauses,
+                openBraceToken: default, members: members, closeBraceToken: default, semicolonToken: default);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs:2849]: It doesn't look like these defaults are going to result in a valid node.

File: src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs [hunk 4]
```diff
@@ -2840,15 +2840,27 @@ public static ConversionOperatorMemberCrefSyntax ConversionOperatorMemberCref(Sy
...
+
+        /// <summary>Creates a new ExtensionDeclarationSyntax instance.</summary>
+        public static ExtensionDeclarationSyntax ExtensionDeclaration(SyntaxToken implicitOrExplicitKeyword, SyntaxToken identifier)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs:2852]: Would it be possible to not add this API in a merge PR?

File: src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs [hunk 5]
```diff
@@ -2840,15 +2840,27 @@ public static ConversionOperatorMemberCrefSyntax ConversionOperatorMemberCref(Sy
...
+            => SyntaxFactory.ExtensionDeclaration(attributeLists: default, modifiers: default,
+                implicitOrExplicitKeyword, keyword: SyntaxFactory.Token(SyntaxKind.ExtensionKeyword), identifier,
+                typeParameterList: null, parameterList: null, forUnderlyingType: null, baseList: null, constraintClauses: default,
+                openBraceToken: default, members: default, closeBraceToken: default, semicolonToken: default);
+
+        /// <summary>Creates a new ExtensionDeclarationSyntax instance.</summary>
+        public static ExtensionDeclarationSyntax ExtensionDeclaration(SyntaxToken implicitOrExplicitKeyword, string identifier)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs:2859]: Would it be possible to not add this API in a merge PR?

File: src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs [hunk 6]
```diff
@@ -2840,15 +2840,27 @@ public static ConversionOperatorMemberCrefSyntax ConversionOperatorMemberCref(Sy
...
-            => SyntaxFactory.ExtensionDeclaration(attributeLists, modifiers, implicitOrExplicitKeyword, keyword, identifier, typeParameterList,
-                forUnderlyingType, baseList, constraintClauses,
-                SyntaxFactory.Token(SyntaxKind.OpenBraceToken), members, SyntaxFactory.Token(SyntaxKind.CloseBraceToken), semicolonToken: default);
-||||||| e9f1cd29802
-=======
+        public static ExtensionDeclarationSyntax ExtensionDeclaration(SyntaxList<AttributeListSyntax> attributeLists, SyntaxTokenList modifiers, SyntaxToken implicitOrExplicitKeyword, SyntaxToken identifier, TypeParameterListSyntax? typeParameterList, ParameterListSyntax? parameterList, ForTypeSyntax? forUnderlyingType, BaseListSyntax? baseList, SyntaxList<TypeParameterConstraintClauseSyntax> constraintClauses, SyntaxList<MemberDeclarationSyntax> members)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs:2845]: I think it would be preferred to not make changes to the shape of this public API in a merge PR

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs]: I took a stab at a more minimal change (new final commit after force push is https://github.com/dotnet/roslyn/commit/2b0adb74b7ccbf5d2905d283c376f6cdb4c2711f). I'm able to avoid any test changes I can't avoid public API changes (because of unavoidable `syntax.xml` change), but I can shrink it a bit (avoid optional `SyntaxFactory.cs` change).

**Review by 333fred — APPROVED** [review]

**Review by AlekseyTs — APPROVED** [review]
LGTM
